R/Gargle-class.R

```diff
@@ -0,0 +1,213 @@
+    config_init = list(),
+    client_credentials = FALSE                  # hard-wired, ok?
+  )
+  Gargle2.0$new(
+    email = email,
+    endpoint = httr::oauth_endpoints("google"), # hard-wired, ok?
+    app = app,
+    params = params,
+    credentials = credentials,
+    cache_path = if (is.null(credentials)) cache else FALSE
+  )
+}
+
+#' OAuth2 token objects specific to Google APIs
+#'
+#' This is based on the [Token2.0][httr::Token] class provided in httr. These
+#' objects should be created through the constructor function
+#' [gargle2.0_token()]. In the base Token2.0 class, tokens are cached based on
+#' endpoint, app, and scopes. For this subclass, the identifier or key is
+#' expanded to include the email address associated with the token. This makes
+#' it easier to work with Google APIs using multiple accounts.
+#'
+#' @docType class
+#' @keywords internal
+#' @format An R6 class object.
+#' @export
+#' @name Gargle-class
+Gargle2.0 <- R6::R6Class("Gargle2.0", inherit = httr::Token2.0, list(
+  email = NA_character_,
+  initialize = function(email = NULL,
+                        endpoint =  httr::oauth_endpoints("google"),
+                        app = gargle_app(),
+                        credentials = NULL,
+                        params = list(),
+                        cache_path = getOption("httr_oauth_cache")) {
+    "!DEBUG Gargle2.0 initialize"
+    stopifnot(
+      httr:::is.oauth_endpoint(endpoint) || !is.null(credentials),
+      httr:::is.oauth_app(app),
+      is.list(params)
+    )
+
+    self$email <- email
+    self$app <- app
+    self$endpoint <- endpoint
+    params$scope <- add_email_scope(params$scope)
+    self$params <- params
+    self$cache_path <- httr:::use_cache(cache_path)
+
+    if (!is.null(credentials)) {
+      # Use credentials created elsewhere - usually for tests
+      self$credentials <- credentials
+      return(self)
+    }
+
+    # Are credentials cached already?
+    if (self$load_from_cache()) {
+      self
+    } else {
+      "!DEBUG no matching token in the cache"
+      self$init_credentials()
+      self$email <- get_email(self) %||% NA_character_
+      self$cache()
+    }
+  },
+  print = function(...) {
+    cat("<Token (via gargle)>\n", sep = "")
+    # print(self$endpoint) ## this is TMI IMO
+    # also, it's boring --> always google for us
+    cat("  <oauth_endpoint> google\n", sep = "")
+    # print(self$app) ## this is TMI IMO
+    cat("       <oauth_app> ", self$app$appname, "\n", sep = "")
+    cat("           <email> ", self$email, "\n", sep = "")
+    ## TODO(jennybc) turn this into a helper or sg
+    scopes <- gsub("/$", "", gsub("(.*)/(.+$)", "...\\2", self$params$scope))
+    cat(
+      "          <scopes> ",
+      paste0(scopes, collapse = ", "),
+      "\n", sep = ""
+    )
+    cat(
+      "     <credentials> ",
+      paste0(names(self$credentials), collapse = ", "),
+      "\n", sep = ""
+    )
+    cat(
+      "                   (expires in ",
```

Comment:
Given that code will autorefresh, should we just elide the expiration?

In particular, I feel like this info is relevant for one of *us*, but likely noise to the *user*.

Reply:
I think you're right long-term. I sense this print method will change even more before we're done. I'm going to leave it for now because I find it useful re: sanity checking that things I think are happening (or not) are actually happening (or not).